libexec/rtld-elf/rtld.c

```diff
@@ -6282,6 +6304,106 @@ _rtld_free_tls(void *tcb, size_t tcbsize, size_t tcbalign)
     lock_release(rtld_bind_lock, &lockstate);
 }
 
+#ifdef CHERI_LIB_C18N
+static bool
+validate_c18nstrtab(const Obj_Entry *obj)
+{
+	const Elf_Phdr *ph;
+
+	if (obj->c18nstrtab != NULL) {
+		if (obj->c18nstrsize == 0)
+			return (false);
+		if (obj->c18nstrtab[obj->c18nstrsize - 1] != '\0')
+			return (false);
+	}
+
+	for (ph = obj->phdr;  (const char *)ph < (const char *)obj->phdr +
+	    obj->phsize; ph++) {
+		switch (ph->p_type) {
+		case PT_COMPARTMENT:
+			if (obj->c18nstrtab == NULL ||
+			    ph->p_paddr >= obj->c18nstrsize)
+				return (false);
+			break;
+		}
+	}
+	return (true);
+}
+
+static void
+c18n_setup_compartments(Obj_Entry *obj, const char *name)
+{
+	Compart_Entry *compart;
+	const Elf_Phdr *ph;
+	size_t len;
+
+	assert(obj->compart_id == 0);
+	obj->compart_id = compart_id_allocate(name);
```

Comment:
It's a bit confusing that both Obj_Entry and Compart_Entry have a compart_id field, and is just asking for someone to accidentally use the Obj_Entry one

Reply:
I don't necessarily disagree.  Having a `Compart_Entry` for the default compartment doesn't quite work though either (its bounds would be the object's entire address range).  In theory, nothing should really be using `obj->compart_id` directly but using the wrapper function to lookup based on address.  I could maybe obfuscate the field name to `obj->default_compart_id`?